core/src/dbs/group.rs

```diff
@@ -0,0 +1,377 @@
+use crate::ctx::Context;
+use crate::dbs::plan::Explanation;
+use crate::dbs::store::StoreCollector;
+use crate::dbs::{Options, Statement, Transaction};
+use crate::err::Error;
+use crate::sql::function::OptimisedAggregate;
+use crate::sql::value::{TryAdd, TryDiv, Value};
+use crate::sql::{Array, Field, Idiom};
+use std::borrow::Cow;
+use std::collections::{BTreeMap, HashMap};
+use std::mem;
+
+pub(super) struct GroupsCollector {
+	base: Vec<Aggregator>,
+	idioms: Vec<Idiom>,
+	grp: BTreeMap<Array, Vec<Aggregator>>,
+}
+
+#[derive(Default)]
+struct Aggregator {
+	array: Option<Array>,
+	first_val: Option<Value>,
+	count: Option<usize>,
+	math_max: Option<Value>,
+	math_min: Option<Value>,
+	math_sum: Option<Value>,
+	math_mean: Option<(Value, usize)>,
+	time_max: Option<Value>,
+	time_min: Option<Value>,
+}
+
+impl GroupsCollector {
+	pub(super) fn new(stm: &Statement<'_>) -> Self {
+		let mut idioms_agr: HashMap<Idiom, Aggregator> = HashMap::new();
+		if let Some(fields) = stm.expr() {
+			for field in fields.other() {
+				if let Field::Single {
+					expr,
+					alias,
+				} = field
+				{
+					let idiom = alias.as_ref().cloned().unwrap_or_else(|| expr.to_idiom());
+					if let Some(agr) = idioms_agr.get_mut(&idiom) {
+						agr.prepare(expr);
+					} else {
+						let mut agr = Aggregator::default();
+						agr.prepare(expr);
+						idioms_agr.insert(idiom, agr);
+					}
```

Comment:
This could be slightly faster by avoiding the second lookup using entry.
```suggestion
					idioms_agr.entry(idiom).or_default().prepare(expr);
```